### PR TITLE
Check if LDSFLD accesses non-static field

### PR DESCRIPTION
**Pull request by sejongoh**

### CODE REVIEW

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -11667,6 +11667,13 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
             }
 
             eeGetFieldInfo(&resolvedToken, (CORINFO_ACCESS_FLAGS)aflags, &fieldInfo);
+
+            // Check if LDSFLD tries to load a non-static field
+            if (opcode == CEE_LDSFLD && !(fieldInfo.fieldFlags & CORINFO_FLG_FIELD_STATIC))
```
Comment (erozenfeld) [inline src/jit/importer.cpp]: cc: There should be parentheses around all unary and binary expressions if they are contained within other expressions.

**Comment by erozenfeld** [comment]
LGTM with one cc nit

**Comment by sejongoh** [comment]
@erozenfeld Thanks for review. I have added parentheses.

**Comment by cmckinsey** [comment]
Does this behavior match JIT32? Do we need to make a patch in that importer too?

File: src/jit/importer.cpp [hunk 2]
```diff
@@ -11669,7 +11669,7 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
             eeGetFieldInfo(&resolvedToken, (CORINFO_ACCESS_FLAGS)aflags, &fieldInfo);
 
             // Check if LDSFLD tries to load a non-static field
-            if (opcode == CEE_LDSFLD && !(fieldInfo.fieldFlags & CORINFO_FLG_FIELD_STATIC))
+            if ((opcode == CEE_LDSFLD) && !(fieldInfo.fieldFlags & CORINFO_FLG_FIELD_STATIC))
```
Comment (RussKeldorph) [inline src/jit/importer.cpp]: Still need more parens around the unary (!) expression.

**Comment by RussKeldorph** [comment]
Please add more details to the description field, such as the facts that this is hardening the JIT against invalid IR and that it addresses #2405.

Also, you should be re-enabling failing tests via tests/*.targets and tests/tests*OutsideWindows.txt files with this change.  If you need to fix tests in order for them to pass since verification isn't enabled in coreclr, please do that, too.  In general, don't leave corresponding test cleanup for later if possible.

File: src/jit/importer.cpp [hunk 3]
```diff
@@ -11667,6 +11667,13 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
             }
 
             eeGetFieldInfo(&resolvedToken, (CORINFO_ACCESS_FLAGS)aflags, &fieldInfo);
+
+            // Check if LDSFLD tries to load a non-static field
+            if ((opcode == CEE_LDSFLD) && !(fieldInfo.fieldFlags & CORINFO_FLG_FIELD_STATIC))
```
Comment (RussKeldorph) [inline src/jit/importer.cpp]: (Repeating comment on non-outdated diff)
Still need more parens around the unary (!) expression.

**Comment by RussKeldorph** [comment]
@cmckinsey Yes, this fails for JIT32 as well.  Will port the VSO bug to dotnet/coreclr.

**Comment by sejongoh** [comment]
I have the same changes for JIT32. I am wondering how to enable the tests for JIT32.

**Comment by RussKeldorph** [comment]
@sejongoh If you remove this test from all of the exclusion files, it will be enabled for JIT32.  Full automated testing for that is still being worked out, but for now that should be all that is required, assuming you make it pass for x64 RyuJIT and x86 legacy backend with this change.

File: src/jit/importer.cpp [hunk 4]
```diff
@@ -11667,6 +11667,13 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
             }
 
             eeGetFieldInfo(&resolvedToken, (CORINFO_ACCESS_FLAGS)aflags, &fieldInfo);
+
+            // Check if LDSFLD tries to load a non-static field
+            if ((opcode == CEE_LDSFLD) && (!(fieldInfo.fieldFlags & CORINFO_FLG_FIELD_STATIC)))
```
Comment (sivarv) [inline src/jit/importer.cpp]: Don't we also need check for CEE_LDSFLDA?

**Comment by sivarv** [comment]
I remember @schellap attempted a similar fix  in the past but that resulted in a regression due to the fix on Desktop.

For details see Fix changeset#1505600 and corresponding rollback changeset#1506481 on Desktop.

**Comment by RussKeldorph** [comment]
@sivarv Did you mean original fix=C1503640, rollback=C1505600?

@schellap you wrote: "Rolling back as the testcase passes with a different CAS policy i.e., throws verification exception and fails only if run directly. Need to investigate further."  Do you have any more details that you can share?  Just trying to avoid re-discovering what you knew in the past.

**Comment by sivarv** [comment]
@RussKeldorph - yes those are correct CS numbers.   I seems to have mistyped

**Comment by schellap** [comment]
This test AV'ed only when running by hand. As part of the automation it always passed, the fix I made passed the hand run situation but broke the automation by throwing and at that time this test wasn't going to meet the bar, so we punted.

JIT SH on desktop should have this test: if you ported the fix to one of the rel/servicing branches on Desktop and ran JIT SH, you'll see failures.

**Comment by RussKeldorph** [comment]
@schellap The Desktop automation enables IL Verification before running the test, and the test case only passes if it receives a VerificationException.  My guess would be that your fix in C1503640 raised an exception that pre-empted the VerificationException, resulting in a test failure.  Does that sound right?  If so, presumably the right fix would be to simply ensure that the VerificationException is thrown when Verification is enabled (Desktop only?) and raise the "BADCODE" exception otherwise.  We'd then change the test case to expect the appropriate exception.

**Comment by schellap** [comment]
@RussKeldorph ah, yes, that sounds correct and the fix sounds good.

**Comment by sejongoh** [comment]
Do we have something like a flag to check if verification is enabled or not? 

I updated the testcase to accept InvalidProgramException from BADCODE(...).

```
catch [mscorlib]System.Security.VerificationException
{
leave.s PASS
}
catch [mscorlib]System.InvalidProgramException
{
leave.s PASS
}
```

**Comment by sejongoh** [comment]
@RussKeldorph On Desktop CLR, tiVerificationNeeded is true and verVerifyField throws a verification exception. On CoreCLR, tiVerificationNeeded is false and it crashes later. The Desktop CLR also crahes in the same place if CAS policy is not properly set up and tiVerificationNeeded is off.

```
// tiVerificationNeeded is turned on for Desktop.
if (tiVerificationNeeded) {
...
  // Raise SEH_VERIFICATION_EXCEPTION exception
  verVerifyField(&resolvedToken, fieldInfo, tiObj, isLoadAddress, bAllowPlainValueTypeAsThis);
...
}

...
...
// CoreCLR crashes here due to null obj.
obj = impCheckForNullPointer(obj);
```

If tiVerificationNeeded indicates whether if verification is enabled or not, we may want to throw an InvalidProgramException when tiVerificationNeeded is off.

**Comment by sejongoh** [comment]
@erozenfeld @RussKeldorph @sivarv PTAL

**Comment by sivarv** [comment]
Looks Good.

File: src/jit/importer.cpp [hunk 5]
```diff
@@ -12055,6 +12067,13 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
                 typeInfo fieldType = verMakeTypeInfo(ciType, fieldClsHnd);
                 Verify(tiCompatibleWith(tiVal, fieldType.NormaliseForStack(), true), "type mismatch");
             }
+            else {
```
Comment (RussKeldorph) [inline src/jit/importer.cpp]: Pretty sure the open-brace belongs on next line.

File: src/jit/importer.cpp [hunk 6]
```diff
@@ -11760,6 +11762,14 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
                     }
                 }
             }
+            else {
```
Comment (RussKeldorph) [inline src/jit/importer.cpp]: Pretty sure open-brace belongs on next line.

**Comment by RussKeldorph** [comment]
LGTM if you fix formatting.

**Comment by sejongoh** [comment]
Thanks for review. I fixed the formatting issue.

**Comment by erozenfeld** [comment]
LGTM

**Comment by sejongoh** [comment]
@dotnet-bot test Windows_NT x86 Release Legacy Backend Build and Test

**Comment by sejongoh** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Pull request merged**